mava/systems/ppo/sebulba/ff_ippo.py

```diff
@@ -11,3 +11,734 @@
+from mava.utils.jax_utils import merge_leading_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.sebulba_utils import ParamsSource, Pipeline, RecordTimeTo, ThreadLifetime
+from mava.utils.total_timestep_checker import check_total_timesteps
+from mava.utils.training import make_learning_rate
+from mava.wrappers.episode_metrics import get_final_step_metrics
+
+
+def rollout(
+    key: chex.PRNGKey,
+    config: DictConfig,
+    rollout_queue: Pipeline,
+    params_source: ParamsSource,
+    apply_fns: Tuple[ActorApply, CriticApply],
+    actor_device_id: int,
+    seeds: List[int],
+    thread_lifetime: ThreadLifetime,
+) -> None:
+    """Runs rollouts to collect trajectories from the environment.
+
+    Args:
+        key (chex.PRNGKey): The PRNGkey.
+        config (DictConfig): Configuration settings for the environment and rollout.
+        rollout_queue (Pipeline): Queue for sending collected rollouts.
+        params_source (ParamsSource): Source for fetching the latest network parameters.
+        apply_fns (Tuple): Functions for running the actor and critic networks.
+        actor_device_id (int): Actor device id for the current thread.
+        seeds (List[int]): Seeds for initializing the environment.
+        thread_lifetime (ThreadLifetime): Manages the thread's lifecycle.
+
+    Returns:
+        None: This function updates the rollout queue with collected data.
+    """
+    # setup
+    env = environments.make_gym_env(config, config.arch.num_envs)
+    current_actor_device = jax.devices()[actor_device_id]
+    actor_apply_fn, critic_apply_fn = apply_fns
+    num_agents, num_envs = config.system.num_agents, config.arch.num_envs
+
+    # Define the util functions: select action function and prepare data to share it with learner.
+    @jax.jit
+    def get_action_and_value(
+        params: FrozenDict,
+        observation: Observation,
+        key: chex.PRNGKey,
+    ) -> Tuple:
+        """Get action and value."""
+        key, subkey = jax.random.split(key)
+
+        actor_policy = actor_apply_fn(params.actor_params, observation)
+        action = actor_policy.sample(seed=subkey)
+        log_prob = actor_policy.log_prob(action)
+
+        value = critic_apply_fn(params.critic_params, observation).squeeze()
+        return action, log_prob, value, key
+
+    timestep = env.reset(seed=seeds)
+
+    next_dones = jnp.repeat(timestep.last(), num_agents).reshape(num_envs, -1)
+
+    move_to_device = lambda x: jax.device_put(x, device=current_actor_device)
+
+    # Loop till the desired num_updates is reached.
+    while not thread_lifetime.should_stop():
+        # Rollout
+        traj: List = []
+        time_dict: Dict[str, List[float]] = {
+            "single_rollout_time": [],
+            "env_step_time": [],
+            "getting_params_time": [],
+            "putting_rollout_time": [],
+        }
+
+        # Loop over the rollout length
+        with RecordTimeTo(time_dict["single_rollout_time"]):
```

Comment:
Can we remove this, I don't think we need the full episode time

Reply:
it's used to track the actor SPS